File: app/src/main/java/br/com/nicetuber/feature/channel_details/ChannelDetailsInteractor.java
package br.com.nicetuber.feature.channel_details;

import com.google.gson.Gson;

import javax.inject.Inject;

import br.com.nicetuber.base.BaseInteractor;
import br.com.nicetuber.util.JsonDeserializer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by thiagozg on 15/10/2017.
 */

public class ChannelDetailsInteractor extends BaseInteractor {

    @Inject
    Gson gson;

    @Inject
    public ChannelDetailsInteractor() {
    }

    public void getChannelStatistics(IChannelDetails.Callback callback, String channelId) {
        youtubeApi.getChannelStatistics(channelId)
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(
                    response -> callback.getChannelStatisticsSuccess(
                            JsonDeserializer.deserializeChannelStatistics(gson, response)),
                    throwable -> {
                        callback.getChannelStatisticsError(throwable.getMessage());
                        throwable.printStackTrace();
                    }
            );
    }
}

File: app/src/main/java/br/com/nicetuber/feature/channel_details/ChannelDetailsInteractor.kt
package br.com.nicetuber.feature.channel_details

import br.com.nicetuber.base.BaseInteractor
import br.com.nicetuber.util.deserializeToChannelStatistics
import com.google.gson.Gson
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import javax.inject.Inject

/**
 * Created by thiagozg on 15/10/2017.
 */

class ChannelDetailsInteractor @Inject
constructor() : BaseInteractor() {

    @Inject
    lateinit var gson: Gson

    fun getChannelStatistics(callback: IChannelDetails.Callback, channelId: String) {
        youtubeApi.getChannelStatistics(channelId)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        { response -> callback.getChannelStatisticsSuccess(
                                        response.deserializeToChannelStatistics(gson))
                        }, {
                            throwable -> callback.getChannelStatisticsError(throwable.message!!)
                                            throwable.printStackTrace()
                        })
    }
}
